### PR TITLE
Evaluate print in evaluation environment

### LINKED ISSUE #192: Special handling for `is.null(output_handler$text)`
Want to consider once last time whether or not it's worth evaluating `print()` in `envir` so it can find any newly defined S3 methods.

### PR DESCRIPTION
**Pull request by hadley**
Fixes #192

This turned out to be rather straightforward to implement after all the refactoring, and while it's unlikely to affect much user code, I think the changes to the tests demonstrates its utility.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # evaluate (development version)
 
+* The default `value` handler now evaluates print in a child environment of the evaluation environment. This largely makes evaluate easier to test, but should make defining S3 methods for print a little easier (@)
```
[the remark this answers is not in the document]
Comment (lionel-) [reply NEWS.md:3]: Maybe:

```suggestion
* The default `value` handler now evaluates print in a child environment of the evaluation environment. This largely makes evaluate easier to test, but should make defining S3 methods for print a little easier (#192).
```

File: R/output.R [hunk 2]
```diff
@@ -23,15 +23,31 @@ show_value <- function(handler, visible) {
...
+  }
+
+  if (isS4(value)) {
+    methods::show(value)
+  } else {
+    # We need to the print() generic in a child environment of the evaluation 
+    # frame in order to find any methods registered there
+    print_env <- new.env(parent = envir) 
+    print_env$value <- value
+    evalq(print(value), envir = print_env)
```
Comment (lionel-) [inline R/output.R:51]: 👍

**Review by lionel- — APPROVED** [review]

**Comment by hadley** [comment]
Ah the problem is that knitr does this:

```
normal_print = default_handlers$value
formals(normal_print) = alist(x = , ... = )
```

**Comment by hadley** [comment]
I made https://github.com/yihui/knitr/pull/2353 so we can eventually drop this extra code.

**Review by cderv — APPROVED** [review]

**Pull request merged**